### PR TITLE
8726 workflow add a test button in workflow code step

### LINKED ISSUE #8726: [Workflow] Add a Test button in workflow code step
- Figma -> https://www.figma.com/design/xt8O9mFeLl46C5InWwoMrN/Twenty?node-id=49745-24459&node-type=frame&m=dev
Image
- Use that to infer code step outputSchema
To do in another ticket 
- Update dropdown variable when no outputSchema is available
Image

### PR DESCRIPTION
**Pull request by martmull**
- add test button to workflow code step
- add test tab to workflow code step

https://github.com/user-attachments/assets/e180a827-7321-49a2-8026-88490c557da2

image
image

### CODE REVIEW

File: packages/twenty-front/src/modules/action-menu/components/CmdEnterActionButton.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  );
+
+  return (
+    <Button
+      title={title}
+      variant="primary"
+      accent="blue"
+      size="medium"
+      onClick={onClick}
+      shortcut={'⌘⏎'}
```
[the remark this answers is not in the document]
Comment (martmull) [reply packages/twenty-front/src/modules/action-menu/components/CmdEnterActionButton.tsx:27]: @Bonapara do we already have a distinction between macOS or Windows/Linux plateform somewhere?

File: packages/twenty-front/src/modules/serverless-functions/constants/IndexFilePath.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { join } from 'path';
```
Comment (Devessier) [inline packages/twenty-front/src/modules/serverless-functions/constants/IndexFilePath.ts:1]: The path module comes from Node.js and shouldn't be used in the browser. I think it will work because Vite polyfills these modules, but it should be avoided.

If you want to manipulate URLs, you can use the [`URL` constructor](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL), or directly concatenate both parts as I think Windows support is not a top-priority.

See https://vite.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility.

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      }));
+    } catch (err) {
+      enqueueSnackBar(
+        (err as Error)?.message || 'An error occurred while executing function',
+        {
+          variant: SnackBarVariant.Error,
+        },
+      );
+    }
+  };
```
Comment (Devessier) [inline packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:60]: Later, we'll have to prevent users from clicking these buttons multiple times while an operation is pending. We'll have to do it in many places to avoid unwanted behaviors.

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                result?.data?.executeOneServerlessFunction?.error,
+                null,
+                4,
+              )
+            : undefined,
+        },
+      }));
+    } catch (err) {
+      enqueueSnackBar(
+        (err as Error)?.message || 'An error occurred while executing function',
```
Comment (Devessier) [inline packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:54]: If think it would be safer to check whether `err` is an `Error`instance.

```suggestion
        err instanceof Error ? error.message : 'An error occurred while executing function',

```

File: packages/twenty-front/src/modules/serverless-functions/utils/getDefaultFunctionInputFromInputSchema.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+          result[key] = getDefaultFunctionInputFromInputSchema([val])[0];
+        });
+      }
+      return result;
+    } else if (param.type === 'array' && isDefined(param.items)) {
+      return [];
+    }
+    return null;
+  });
+};
```
Comment (Devessier) [inline packages/twenty-front/src/modules/serverless-functions/utils/getDefaultFunctionInputFromInputSchema.ts:24]: 👍 👌

File: packages/twenty-front/src/modules/serverless-functions/utils/getFunctionInputFromSourceCode.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import { getDefaultFunctionInputFromInputSchema } from '@/serverless-functions/utils/getDefaultFunctionInputFromInputSchema';
+import { FunctionInput } from '@/workflow/types/FunctionInput';
+import { isObject } from '@sniptt/guards';
+import { isDefined } from 'twenty-ui';
+
+export const getFunctionInputFromSourceCode = (
+  sourceCode?: string,
+): FunctionInput => {
+  if (!isDefined(sourceCode)) {
+    throw Error('Source code is not defined');
```
Comment (Devessier) [inline packages/twenty-front/src/modules/serverless-functions/utils/getFunctionInputFromSourceCode.ts:11]: Calling `Error()` returns a new `Error` instance, but it would be more idiomatic to call `throw new Error()`. For some people, this is not a better pattern. However, it is most commonly used across the codebase.

File: packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormDeleteRecord.tsx [hunk 7]
```diff
@@ -118,52 +119,55 @@ export const WorkflowEditActionFormDeleteRecord = ({
...
-          const newFormData: DeleteRecordFormData = {
-            objectName,
-            objectRecordId: '',
-          };
-
-          setFormData(newFormData);
-
-          saveAction(newFormData);
+    <>
+      <WorkflowStepHeader
```
Comment (Devessier) [inline packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormDeleteRecord.tsx:123]: I like the new API 👍 It embraces composition better.

File: packages/twenty-front/src/modules/ui/layout/tab/components/TabList.tsx [hunk 8]
```diff
@@ -30,13 +29,17 @@ type TabListProps = {
...
 
-const StyledContainer = styled.div`
-  border-bottom: ${({ theme }) => `1px solid ${theme.border.color.light}`};
+const StyledTabsContainer = styled.div`
   box-sizing: border-box;
   display: flex;
   gap: ${({ theme }) => theme.spacing(2)};
   height: 40px;
   user-select: none;
+  margin-bottom: -1px;
```
[the remark this answers is not in the document]
Comment (martmull) [reply packages/twenty-front/src/modules/ui/layout/tab/components/TabList.tsx:38]: I don't have any other idea to produce that design without negative margin. Maybe you do?
image

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts (hunk 4 above)
Comment (Devessier) [reply packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:54]: If we take a step back, I wonder if displaying programmatic errors to the user is excellent. We can see later if we can improve it.

Comment (Devessier) [reply packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:54]: Please also note that the condition `err instanceof Error` is likely ever to be true.

**Review by Devessier — COMMENTED** [review]
Great job! I only gave a look to the first half of the Pull Request. I wrote a few comments.

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts (hunk 4 above)
Comment (martmull) [reply packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:54]: Nice catch, that's probably a relicate of old bad behavior, will improve that

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts (hunk 3 above)
Comment (martmull) [reply packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:60]: Updating or creating a serverless function is almost instant now. The serverless lambda build and deployment which can take several seconds, is now handled asynchronously with a job
The only weird bahavior that can remain is that when you test your function just after updating it, the result my be the result of the previous version during the deployment. It will be address with a while loop that check if function is properly deployed

File: packages/twenty-front/src/modules/serverless-functions/utils/getFunctionInputFromSourceCode.ts (hunk 6 above)
Comment (martmull) [reply packages/twenty-front/src/modules/serverless-functions/utils/getFunctionInputFromSourceCode.ts:11]: yes that's a typo thanks!

Comment (thomtrp) [reply packages/twenty-front/src/modules/serverless-functions/utils/getFunctionInputFromSourceCode.ts]: @martmull can we throw a specific error for Serverless there?

File: packages/twenty-server/src/modules/workflow/workflow-builder/types/output-schema.type.ts [hunk 9]
```diff
@@ -24,3 +24,9 @@ export type RecordOutputSchema = {
 };
 
 export type OutputSchema = BaseOutputSchema | RecordOutputSchema;
+
+export type StepOutputSchema = {
+  id: string;
+  name: string;
+  outputSchema: OutputSchema;
+};
```
Comment (thomtrp) [inline packages/twenty-server/src/modules/workflow/workflow-builder/types/output-schema.type.ts:32]: For what is it used? 🤔

File: packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    });
+
+    if (isDefined(result?.data?.executeOneServerlessFunction?.data)) {
+      callback?.(result?.data?.executeOneServerlessFunction?.data);
+    }
+
+    setServerlessFunctionTestData((prev) => ({
+      ...prev,
+      language: 'json',
+      height: 300,
```
[the remark this answers is not in the document]
Comment (martmull) [reply packages/twenty-front/src/modules/serverless-functions/hooks/useTestServerlessFunction.ts:28]: Because we init the height small when no result exists

image

File: packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormServerlessFunction.tsx [hunk 11]
```diff
@@ -89,155 +87,123 @@ export const WorkflowEditActionFormServerlessFunction = ({
...
-        {
-          variant: SnackBarVariant.Error,
-        },
-      );
+  const updateOutputSchemaFromTestResult = async (testResult: object) => {
+    if (actionOptions.readonly === true) {
+      return;
     }
+    const newOutputSchema = getFunctionOutputSchema(testResult);
+    await updateAction({
```
Comment (Devessier) [inline packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormServerlessFunction.tsx:95]: `await` is unnecessary. The `updateAction` doesn't return a `Promise`.

File: packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormServerlessFunction.tsx [hunk 12]
```diff
@@ -89,155 +87,123 @@ export const WorkflowEditActionFormServerlessFunction = ({
...
   };
 
   const handleEditorDidMount = async (
     editor: editor.IStandaloneCodeEditor,
     monaco: Monaco,
   ) => {
+    editor.addCommand(
+      monaco.KeyMod.CtrlCmd | monaco.KeyCode.Enter,
+      handleRunFunction,
+    );
```
Comment (Devessier) [inline packages/twenty-front/src/modules/workflow/workflow-actions/components/WorkflowEditActionFormServerlessFunction.tsx:206]: I think we should remove this command if we remove the button on the bottom.

**Review by Devessier — CHANGES_REQUESTED** [review]
Great work! Thank you!
As discussed, removing the "Test" button from the "Code" would be great. I think it doesn't provide the best UX as the code and the test parts are separate, making it hard to know the arguments used when launching a test from the "Code" tab.

File: nx.json [hunk 13]
```diff
@@ -112,7 +112,7 @@
       "outputs": ["{projectRoot}/{options.output-dir}"],
       "options": {
         "cwd": "{projectRoot}",
-        "command": "VITE_DISABLE_ESLINT_CHECKER=true storybook build",
+        "command": "VITE_DISABLE_TYPESCRIPT_CHECKER=true VITE_DISABLE_ESLINT_CHECKER=true storybook build",
```
Comment (Devessier) [inline nx.json:115]: What is the reason behind this change?

**Review by Devessier — APPROVED** [review]
Amazing! Approving.

File: nx.json (hunk 13 above)
Comment (martmull) [reply nx.json:115]: a discussion about performances with @Charles. Typescript was run twice, this change make it run once
